python_legacy/iceberg/api/partition_spec.py

```diff
@@ -76,7 +76,7 @@ def partition_type(self):
             result_type = field.transform.get_result_type(source_type)
             struct_fields.append(NestedField.optional(field.field_id,
                                                       field.name,
-                                                      result_type))
+                                                      result_type.get()))
```

Comment:
I'll admit that I thought all types had a `get()` method...  I only tested it with my use-case, which had a StringType.  I don't mind turning my PR into a bigger fix, but I'll wait for your advice 😄

Reply:
I looked at this a bit more. What's happening is that the `get` method is only included for the types that store the first instance in the `self.__instance` class attribute, presumably to get singleton behavior. This method isn't used for `TimestampType`, `DecimalType`, or `FixedType` since each instance could be different (w or w/o timezone, precision & scale, and length, respectively). This [test](https://github.com/apache/iceberg/blob/master/python_legacy/tests/api/test_partition_spec.py#L36) shows what I'm describing.

I think instead of calling `get()` after the fact, we need to fix the particular `get_result_type` implementation that's not returning a type. Were you by any chance using a `truncate` transform for one of the partition fields? That's the only transform I see with a `get_result_type` that's returning the source type's class and not a type instance. [truncate.py#L51](https://github.com/apache/iceberg/blob/master/python_legacy/iceberg/api/transforms/truncate.py#L51)

Reply:
I agree with @samredai. This shouldn't be calling `get`. Instead, `get_result_type` should be returning a type instance instead of a class.

Reply:
@samredai You guessed it right, I was using `truncate` on a string.  I always wondered why this bug had not been caught earlier and now I understand why.  I guess partitioning on a truncated field is not that common after all?!  I have modified the truncate transform and reverted my change to partition_spec.py and ran this code across all my Iceberg tables: did not get any error.  Changing `get_result_type` for truncate didn't trip any test cases, so I assume there is a gap there.  Something to keep in mind for the refactoring 😄

Reply:
> @samredai You guessed it right, I was using `truncate` on a string.  I always wondered why this bug had not been caught earlier and now I understand why.  I guess partitioning on a truncated field is not that common after all?!  I have modified the truncate transform and reverted my change to partition_spec.py and ran this code across all my Iceberg tables: did not get any error.  Changing `get_result_type` for truncate didn't trip any test cases, so I assume there is a gap there.  Something to keep in mind for the refactoring 😄 

Hi sorry for the issues here. I wouldn't  take this as a signal that truncate isn't common. Most of this was written specifically for use cases at Netflix very early in the development of iceberg. This is part of the reason why a redesign is being pursued rather than completing the current implementation. 

Again, apologies  for any issues you ran into related to this.